### DISCUSSION TITLE
Online playground to run & test the exercises

### DISCUSSION THREAD

**Discussion by hatemhosny** [EmNudge/watlings#9] [General]
Hi @EmNudge 

I have just come across your great project. Excellent work!

I'm the author of [LiveCodes](https://livecodes.io/), an [open-source](https://github.com/live-codes/livecodes), [feature-rich](https://livecodes.io/docs/features/), [client-side](https://livecodes.io/docs/why#client-side), code playground that supports [80+ languages/frameworks](https://livecodes.io/docs/why#language-support).

Check: [Introducing LiveCodes](https://blog.livecodes.io/introducing-livecodes)
Documentations: https://livecodes.io/docs/

LiveCodes supports WebAssembly Text Format among the wide range of language support. It uses [wabt.js](https://github.com/AssemblyScript/wabt.js) to run WAT in the browser. The editor uses Monaco editor (that powers VS Code) on the desktop and CodeMirror6 on mobile. These offer very nice syntax highlighting and code editing experience.

A minimal helper code is added to allow interoperability with the page JS for getting access to imports/exports.

Check the starter template: https://livecodes.io/?template=wat

In addition, there is a [test runner](https://livecodes.io/docs/features/tests) that uses Jest (also runs in the browser).

I was able to get your first exercise running (including the tests) without much effort: 
https://livecodes.io/?x=id/ad92hkkraxy&tests=full

LiveCodes supports [embedding playgrounds](https://livecodes.io/docs/features/embeds) in any webpage. In addition, there is an easy-to-use, yet powerful [SDK](https://livecodes.io/docs/sdk) that allows creating embedded playgrounds and communicating with them (e.g. get/set code, run tests and get their results, watch console output, etc).

Check the article: [Let's Make an Interactive Coding Tutorial](https://blog.livecodes.io/lets-make-an-interactive-coding-tutorial)

I thought LiveCodes can be useful for your project, where users can just run code online and check their solutions without having to download or install anything, even on mobile, with syntax highlighting, share/export projects, save for later, and lots of other [features](https://livecodes.io/docs/features/).

If you are interested, I'd be happy to help in creating interactive playgrounds for all your exercises.

**Comment by EmNudge** [comment]
I've added a web-version for this repository now at https://watlings.emnudge.dev
I was waiting for a powerful LSP implementation before offering a hosted version. I've since built it at https://github.com/EmNudge/wat-lsp

You're more than welcome to re-use the exercises here and host them on your playground.
The above LSP is available at `@emnudge/wat-lsp` if you'd like to use it for your editor.
